awx/main/models/workflow.py

```diff
@@ -600,3 +602,56 @@ def actually_running(self):
         # WorkflowJobs don't _actually_ run anything in the dispatcher, so
         # there's no point in asking the dispatcher if it knows about this task
         return self.status == 'running'
+
+
+class WorkflowApprovalTemplate(UnifiedJobTemplate):
```

Comment:
I don't yet understand this feature very well, but my personal preference would be to try and avoid subclassing from the unified job template and unified job models. There is a lot of baggage that come along with those.

Reply:
@AlanCoding 

@chrismeyersfsu and I spoke at length with @beeankha about this and in this case, we've intentionally decided that we *want* a lot of this baggage, because these objects will end up behaving a *lot* like jobs in our system.  By making these nodes just be UJT/UJ implementations, we get a _lot_ for free, such as:

* No need to change the task manager or DAG logic - from the perspective of how a workflow's state machine works, these are _just_ jobs.  When they go into `pending` state, however, we don't actually launch a real playbook under the hood.  The process of transitioning from `pending` to `success` or `failure` isn't bound to some background process - it's determined by manual approval by a user (or some API client).
* We get some common relationships we care about for free (such as notification templates)

Reply:
> Approval templates are _just_ UJTs.

Then use UJTs.

```python
In [1]: UnifiedJobTemplate.objects.create()
Out[1]: <UnifiedJobTemplate: -20>
```

Whenever you introduce a new sub-model, it piggybacks onto a million places that you don't expect, and it introduces things that we will be cleaning up from for a very long time. Django polymorphic is a den of 🐉s, and adding to the sub-types scales very poorly. If we do this with a new sub-model, then we need very through investigation with a lot of data to see what actual queries are being produced while interacting with the unified models.

We will still have problems creating bare UJTs, because some code is sure to make assumptions that these do not exist. But that is the lesser of the two evils without a doubt. The sub-model is adding things on top of an unstable house of cards.

Reply:
> Then use UJTs.

These _seems_ fine until later on down the road when we decide to add behavior to these nodes (i.e., "allow a user to specify a timeout on the approval node").  Then you arrive at a place where you really *do* want a separate table for the polymorphic subclass to hold _its_ particular attributes.

>  If we do this with a new sub-model, then we need very through investigation with a lot of data to see what actual queries are being produced while interacting with the unified models.

That's fair - it's possible that by introducing a new UJT and UJ implementation, we have bleed-through in places (these approval nodes start showing up in places we don't expect).  This is the type of thing I'd _hope_ we would catch through our integration test suite.

Reply:
@AlanCoding another question to this approach.  How would you query something like this?

"List all of the `UnifiedJob`s which represent 'ready to be approved/rejected'."

In the current implementation, this is very straightforward:

`/api/v2/workflow_approvals/?status=pending`

Reply:
Two very good questions.

> "List all of the `UnifiedJobs` which represent ready to be `approved/rejected`."

If I follow my proposal, `/api/v2/unified_jobs/?status=pending&polymorphic_ctype__isnull=true`. That syntax isn't great. If we are honest about the function of the bare UJT, it has nothing to do with approvals, it's a placeholder for the task manager... it might be used in some other place as this function, so this is imperfect. In the `workflow_approvals` you could list the workflow job node instead of the UJ record - just one idea.

> If we treated these as vanilla UJTs, we'd probably end up adding some new `ApprovalConfig` model/foreign key on `WorkflowJobNode` and issuing requests like this:

Maybe? Maybe later, or maybe as a way to more cleanly expose the workflow approvals (from first question).

A related conversation: What if you allowed for multiple approvers? You might add a many-related relationship to the `User` model? No. This is a bad idea. We already have a tool for this - roles. Say we introduce `approval_role`. If we did that, it would probably be on the WFJT model, and apply to all approval nodes in the workflow. Anything else gets kind of crazy.

So we need room for future possible expansion of the feature, but at least some of the bells and whistles that we add would be on the WFJT model. That's why I would like keep the model additions as close to the workflow models as reasonable, as opposed to the unified job models.

Reply:
I think my biggest concern about this approach is that we're _basically_ overloading "/api/v2/unified_jobs/?&polymorphic_ctype__isnull=true" to *actually* represent a new type of UJT, but we're just doing it in a way that's not super obvious or discoverable.

Reply:
Maybe there is something to this notion of an "approval role", though.  If we went this route, `WorkflowJobTemplateNode` could have a new implicit role field.  Perhaps the "show me everything that needs approval" could be something like "show me all UnifiedJobs that are `polymorphic_ctype__isnull=true` and which have one or more approvers"?

Reply:
> it introduces things that we will be cleaning up from for a very long time

@AlanCoding I know we've spoken about this privately, but can you voice some examples of these concerns here?  I want to weigh all sides of this and make sure we've really thought out our approach here.

Reply:
I don't see a problem using the polymorphic system as intended, and I think it's a particularly bad idea to use the unified models without defining a concrete representation which is the whole point of the polymorphic/unified system. The Unified model system was *always* supposed to only be used in the case where you needed to cut a query or logic across *all* job types or list them in a unified way.

The polymorphic system does piggyback off of the idiomatic ContentType system within Django, hasn't caused us a *whole* lot of grief, and is still well maintained after all of these years of our using it. I agree that we have to be careful with performance, but that's fine... if that's the only concrete complaint then that's something we can work with. The complexity of dealing with this using non-polymorphic types would completely dwarf that concern.

Either we continue to embrace the polymorphic types and use them as intended or, if they are causing us more problems than are benefiting us, we need to look into replacing or removing them.

Reply:
There are three ideas on the table.
1. (current implementation) inherit from `UnifiedJobTemplate` + `UnifiedJob`
2. (alan) use `UnifiedJobTemplate` + `UnifiedJob`
3. (else) Some sort of WorkflowNode extension

1. Sure, there might be fallout but remember what we get for free: scheduler just works, notifications easy, UI just works.
2. No.
3. We could explore this in another PR. I like it from a design perspective. I'm worried about the amount of scheduler changes needed from a *performance* aspect and correctness.

I vote for 1.

Reply:
@AlanCoding,

Do you mind itemizing some examples you can think of off-hand so we can make sure to pay attention to this?

> Whenever you introduce a new sub-model, it piggybacks onto a million places that you don't expect, and it introduces things that we will be cleaning up from for a very long time.

Reply:
> Does this mostly stem from the UI potentially needing to update query arguments (places where we list jobs and templates)

I'm not too concerned about templates.  AFAIK the templates lists are already filtered down to specific types: `/api/v2/unified_job_templates/?page_size=20&order_by=name&type=workflow_job_template,job_template` so I actually don't think we'd have to worry about these new templates showing up in the UI where they shouldn't.  I'll make sure to double check that we don't have any unfiltered requests to that endpoint.  

As far as unified jobs go - I guess we need to talk about what our expectations are here.  If we expect jobs with `type=workflow_approval` to be excluded from these list(s) then yes we will need to update our queries.  Right now the jobs list looks like: `/api/v2/unified_jobs/?page_size=20&order_by=-finished&not__launch_type=sync`.  However, exposing these jobs in the list could be one way that an administrator approves/denies the pause node.  If these jobs are not exposed in that list then I'm not sure that I understand how we'll be able to tell if a `workflow_job` is stuck on a pause node and whether or not the logged in user can approve/deny.  That sounds like a bunch of summary_field magic.  If we are planning on exposing all of that on the outer `workflow_job` then that needs to be pretty dynamic.  The `workflow_job` could be waiting on N pause nodes if all of those nodes were defined on disparate branches.

That might have been really confusing...

Let me try to convey with pictures.  Lets say we have a simple workflow that looks like:
<img width="1677" alt="Screen Shot 2019-07-11 at 10 37 50 AM" src="https://user-images.githubusercontent.com/9889020/61060083-0ab60b00-a3c8-11e9-9795-a3bf0f11a082.png">

When that workflow gets launched, two new jobs are created.  The first is type: `workflow_job` and the second (the job for the node) is type `workflow_approval`:
<img width="1437" alt="Screen Shot 2019-07-11 at 10 38 11 AM" src="https://user-images.githubusercontent.com/9889020/61060214-333e0500-a3c8-11e9-997b-9518978a4357.png">

Showing `workflow_approval` jobs may make it easier for the UI to expose a way to approve/deny the job right from the jobs list.  If we don't show those jobs then someone may need to explain to me how we can inspect the `workflow_job` for information around whether this user can approve/deny and what the corresponding endpoints are since the ID will be that of the `workflow_approval` job.  

The `workflow_job` may be waiting on several approval nodes at the same time and I'm curious how that would be handled:
<img width="1680" alt="Screen Shot 2019-07-11 at 10 44 34 AM" src="https://user-images.githubusercontent.com/9889020/61060675-e9a1ea00-a3c8-11e9-8464-e425ecaa4138.png">

<img width="1433" alt="Screen Shot 2019-07-11 at 10 42 58 AM" src="https://user-images.githubusercontent.com/9889020/61060577-bc553c00-a3c8-11e9-8504-efefd916a968.png">

Reply:
> However, exposing these jobs in the list could be one way that an administrator approves/denies the pause node. 

I think I like this idea.

Reply:
Thanks for the screen shots @mabashian, I agree that if the intent is to expose these in the JOBS list, then the new type is unavoidable, because the name alone would not be sufficient for the user to identify it. Just like there's a badge for "workflow job", there would need to be a badge like "approval job".

> When we get to the approval RBAC, perhaps the `WorkflowApprovalTemplate` could have a new approval role that users can belong to 

I thought it was deemed sufficient to use some existing role, such as the WFJT admin_role (maybe an org-level role? this has probably already been asked) for a user to approve. For an MVP, an approval role doesn't seem necessary. I'm also concerned about unanticipated effects adding a role on WFJT nodes, particularly since inheritance is likely to go workflow_job_template->organization->admin_role, and a long chain of editable relationships in parentage has caused problems with JT roles in the past.

Reply:
One open-ended question is how we handle the name.  UJT.name *is* enforced as unique, but unlike projects, job templates, etc... the notion of a name (unique across the org) for one of these doesn't really make much sense. Would we want to auto-generate names for these?  I can't think of any actual name that would make them identifiable, and in fact you wouldn't actually ever *look for* and select on of these templates from a list; by their very nature, they're adhoc and ephemeral.

Reply:
Or maybe we *should* treat these as reusable?  Maybe people might want to make them and re-use them e.g., "Requires Manager Approval")

Reply:
I vote for having these approval nodes be reusable, since there may be instances (as @ryanpetrello points out) where there is a standard, regularly-occuring ✔️that's required by the same person/group and is done repeatedly for certain types of workflows.  They can be uniquely named, just like any job template or workflow, and can be easily identifiable by the user that way.

Just to expand on this point, re-creating a notification every time you want to be notified of something = discouraging, but since notifications are permanent (until deleted) and assignable at will, they're much more useful.  Approval nodes could be thought of in a similar way, usage-wise.

Reply:
Catching up here....I see the usefulness of having reusable ones. Where did y'all end on that conversation?

Reply:
@kdelee current documented design plan is not allowing reuse, and creating them by a related endpoint of WFJT nodes, thereby having them tied to a node.